interfaces: require PPP interface to be in up state

nscheer opened the pull request:
Adds `-u` flag to ifconfig check to ensure only interfaces that are UP are reported. Furthermore adds a check for the output of ifconfig to ensure that the interface in question in fact is listed.

Otherwise the check will "approve" an interface that is not yet UP, and subsequent calls to fetch the gateway of that interface might fail if issued to quickly thereafter.

Resolves #7691

fichtner commented:
@nscheer thanks, safe to say this code wasn't changed in a decade which is good and bad at the same time.changing this can easily break behaviour for others as well. I need to dig into the code history first.

nscheer commented:
@fichtner Thanks for looking into this!
Yes, there might be side effects. Question is if this ever should continue without an interface being UP.
And there still is the 10 second "timeout" when it will continue, no matter what.

Let me know, if I can help or test anything :)

fichtner commented:
Ok so first things first this was Seth in pfSense via https://github.com/pfsense/pfsense/commit/302d646e9c25 and unfortunately his work seldomly mentions tickets or problem scope. The behaviour changed very little except that they now wait around 30 seconds for this to complete which tells us how problematic this implementation is to this day... yes I know it just continues after that time anyway ;)

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1248,11 +1248,12 @@ EOD;
-        exec("/sbin/ifconfig " . escapeshellarg($ppp['if']) . " 2>&1", $out, $ret);
-        if ($ret == 0) {
+        $out = [];
+        exec("/sbin/ifconfig -u " . escapeshellarg($ppp['if']) . " 2>&1", $out, $ret);
+        if ($ret == 0 && strpos(implode("\n", $out), $ppp['if']) !== false) {
```
using aadc34d6a0b this can be:

```suggestion
        if (does_interface_exist($ppp['if'], 'up')) {
```

fichtner commented:
Modernizing the existence check out of the way I don't mind making this wait up to 30 seconds as well. The only thing that I think is worth changing is that when 30 seconds passed without the interface being up a log message should be emitted for traceability?

fichtner commented:
Just a thought: for further consistency we could consider bringing the interface down if it exists before starting mpd5 again?

fichtner commented:
@nscheer is it ok to offload this to you or do you want me to take this over? just because I'd like to get this into the code rather sooner than later.

nscheer commented:
> @nscheer is it ok to offload this to you or do you want me to take this over? just because I'd like to get this into the code rather sooner than later.

Yes, let me update my branch accordingly. Will get back to you, if something’s unclear.
Might be rather late today, tough :)

fichtner commented:
No problem. Let me know if I can help 👍

nscheer commented:
@fichtner I rebased on master and added the minimal changes I thought were needed.
Usually I'd like to mention the threshold value (i.e. 30 seconds) in the log message - but I'd rather introduce a constant for that than having the literals spread all over the place. Didn't know if there where any project specific style guides about that.
Let me know if I should adjust that.

nscheer commented:
Since we are talking about modernizing:
You just added the flag parameter to `does_interface_exist` and it has a default value.
Is it correct that this function is „newer“ than the legacy list function?
What do you think of adding an enum of interface states and require this in `does_interface_exist`? This way, we’d enforce correct use and have it self documenting. The enum can be backed by the string values the legacy function expects.

This would clean up the api a bit and we’d already use the enum for the first time.

fichtner commented:
1. We try to modifications in these parts of the code unless there is a fundamental technical issue that needs fixing. I just adopted the same style as `legacy_interface_listget()` as you can see this code is rarely touched over the years because risks of regressions are real.
2. There is probably very little benefit from sinking time here into a once-used API. Eventually when the MVC/API efforts shift to interface pages this whole code base is in need of a major overhaul anyway. But even then I doubt there is much need of checking if an interface exists and is up with this instance being the first in the code after many years without it.


Cheers,
Franco

nscheer commented:
Agreed, thanks for the clarification!

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1248,17 +1248,22 @@ EOD;
     }
 
+    if (!$interface_is_up) {
+        log_msg("Waiting threshold exceeded - interface $interface is still not UP in interface_ppps_configure().", LOG_WARNING);
```
```suggestion
        log_msg("interface_ppps_configure() waiting threshold exceeded - device {$interface} is still not up", LOG_WARNING);
```

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1248,17 +1248,22 @@ EOD;
     shell_safe('/usr/sbin/daemon -f /usr/local/opnsense/scripts/interfaces/ppp-rename.sh %s %s %s', [$interface, $ifcfg['if'], $ppp['type']]);
 
-    /* wait for up to 10 seconds for the interface to appear (ppp(oe)) */
+    /* wait for up to 30 seconds for the interface to appear in UP state (ppp(oe)) */
```
```suggestion
    /* wait for functional device */
```

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1248,17 +1248,22 @@ EOD;
-    while ($i < 10) {
-        exec("/sbin/ifconfig " . escapeshellarg($ppp['if']) . " 2>&1", $out, $ret);
-        if ($ret == 0) {
+    $interface_is_up = false;
```
```suggestion
```

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1248,17 +1248,21 @@ EOD;
-        if ($ret == 0) {
+    while ($i < 30) {
+        if (does_interface_exist($ppp['if'], 'up')) {
+            $interface_is_up = true;
```
```suggestion
```

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1248,17 +1248,20 @@ EOD;
         $i++;
     }
 
+    if (!$interface_is_up) {
```
```suggestion
    if ($i >= 30) {
```

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1248,17 +1248,20 @@ EOD;
-        exec("/sbin/ifconfig " . escapeshellarg($ppp['if']) . " 2>&1", $out, $ret);
-        if ($ret == 0) {
+    while ($i < 30) {
+        if (does_interface_exist($ppp['if'], 'up')) {
             break;
         }
         sleep(1);
```
```suggestion
        sleep(1);
        if (does_interface_exist($ppp['if'], 'up')) {
            break;
        }
```

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1248,17 +1248,20 @@ EOD;
     shell_safe('/usr/sbin/daemon -f /usr/local/opnsense/scripts/interfaces/ppp-rename.sh %s %s %s', [$interface, $ifcfg['if'], $ppp['type']]);
 
-    /* wait for up to 10 seconds for the interface to appear (ppp(oe)) */
+    /* wait for functional device */
```
```suggestion
    /* wait for functional device */
    $max = 30;
```

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1248,17 +1248,21 @@ EOD;
-    while ($i < 10) {
-        exec("/sbin/ifconfig " . escapeshellarg($ppp['if']) . " 2>&1", $out, $ret);
-        if ($ret == 0) {
+    while ($i < 30) {
```
```suggestion
    while ($i < $max) {
```

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1248,17 +1248,21 @@ EOD;
         $i++;
     }
 
+    if ($i >= 30) {
```
```suggestion
    if ($i >= $max) {
```

fichtner commented on src/etc/inc/interfaces.inc:
```diff
@@ -1239,6 +1239,9 @@ EOD;
     }
 
+    /* precaution for post-start 'up' check */
+    legacy_interface_flags($interface, 'down');
```
```suggestion
    legacy_interface_flags($interface, 'down', false);
```

fichtner commented:
@nscheer merged, thanks! I modified it a bit. Can you make sure it still works as intended?

fichtner commented:
@nscheer FYI https://github.com/opnsense/core/commit/7c9b97ab74c

nscheer commented:
@fichtner Nice! Yes, still works as intended!

fichtner commented:
Thanks, will be in 24.7.2. 😊
